### DISCUSSION TITLE
Tile Editor

### DISCUSSION THREAD

**Discussion by skeddles** [lospec/pixel-editor#79] [Feature Suggestions]
It would obviously be great to have a tile editor, but I have an idea for it that would take it one step further.

### Normal Tile Editors
The base features of a tile editor would be
- define a tile size (and possibly restrict the document to a multiple of this size)
- create a tile set (either add tiles then draw them or create tiles from spaces on canvas)
- place tiles
- editing tiles updates all tiles
- possibly some advanced features like tile layouts which automatically draw borders

### Grid Free Tile Editing
My idea is instead of having a set size for tiles, each tile could be any size, and the tiles don't have to align to the grid. So it would end up being something closer to like a room editor in a game engine, where you just place objects down. Instead of having to create a tree out of 8 tiles, you could just make the tree sprite, and place it around the room. You could easily create a tile out of a layer.

But if you desire you could still create all the tiles at the same resolution, and enable grid snapping to do tilemapping the traditional way (so all the features of a normal tile editor, but more flexible).

While hovering these tiles you would see a blue outline/name, and can click to select it which allows you to place more of it.

### Animation with tiles
This would probably be added at a separate time, but animation is where this type of tile would become extremely powerful: each tile could have it's own looping animation, which would let you easier combine multiple animations into one, simply by layering them on top of each other. This would be similar to how "symbols" worked in adobe flash, each has it's own timeline, which plays in a loop. This way if you wanted to mix two animations at a different framerate, you wouldn't have to create each frame up to the point when it lined up, it would just happen automatically.

Animated symbols could be exported separately as gifs or framestrips for use in a game engine.

In my dream world, you could also create key frames for properties of these tiles on the animation timeline, such as their position, and it would automatically animate between the two (similar to how aftereffects, flash works). This would enable people to make full fledged animations (and not just looping things). There is no good pixel art animation tool at the moment, most people have to resort to game engines or pixel unfriendly tools like after effects.

**Comment by unsettledgames** [comment]
This is interesting. Implementing the first part should be fairly feasible, even though it would require implementing multiple files in the same window for a better workflow when editing tiles imo. An alternative would be to place all the tiles in a different canvas where the user can edit them; the user can decide to load one tile at a time or a whole tileset.

Regarding the second part (animation), I feel it's a bit overscoped. Most of the features you're describing are made in engines and people would probably keep sticking to them: if we wanted to implement keyframes, we'd have to implement a huge number of additional features in order for them to be usable. 

The rest of the animation section could be feasible, but I'd postpone tile editing until we have both multifile editing and a working animation timeline.

Anyway,  a mockup of the interface needed (just for the basic tileset editing) would be great. I'm imagining something that can be hidden and shown only when necessary to avoid creating clutter.

P.S.: after we have a basic tile editor, I'd like to focus on its features before moving on to animation. Tilesets are generally static, so I'd rather focus on static features before moving on to animation (auto tiling would be neat)
